// @datadog/browser-rum/cjs/viewCollection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var browser_core_1 = require("@datadog/browser-core");
var lifeCycle_1 = require("./lifeCycle");
var trackEventCounts_1 = require("./trackEventCounts");
var trackPageActivities_1 = require("./trackPageActivities");
var ViewLoadingType;
(function (ViewLoadingType) {
    ViewLoadingType["INITIAL_LOAD"] = "initial_load";
    ViewLoadingType["ROUTE_CHANGE"] = "route_change";
})(ViewLoadingType = exports.ViewLoadingType || (exports.ViewLoadingType = {}));
exports.THROTTLE_VIEW_UPDATE_PERIOD = 3000;
function startViewCollection(location, lifeCycle, session) {
    var currentLocation = tslib_1.__assign({}, location);
    var startOrigin = 0;
    var currentView = newView(lifeCycle, currentLocation, session, ViewLoadingType.INITIAL_LOAD, startOrigin);
    // Renew view on history changes
    trackHistory(function () {
        if (areDifferentViews(currentLocation, location)) {
            currentLocation = tslib_1.__assign({}, location);
            currentView.end();
            currentView = newView(lifeCycle, currentLocation, session, ViewLoadingType.ROUTE_CHANGE);
        }
    });
    // Renew view on session renewal
    lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.SESSION_RENEWED, function () {
        currentView.end();
        currentView = newView(lifeCycle, currentLocation, session, ViewLoadingType.ROUTE_CHANGE);
    });
    // End the current view on page unload
    lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.BEFORE_UNLOAD, function () {
        currentView.end();
    });
}
exports.startViewCollection = startViewCollection;
function newView(lifeCycle, location, session, loadingType, startOrigin) {
    if (startOrigin === void 0) { startOrigin = performance.now(); }
    // Setup initial values
    var id = browser_core_1.generateUUID();
    var measures = {
        errorCount: 0,
        longTaskCount: 0,
        resourceCount: 0,
        userActionCount: 0,
    };
    var documentVersion = 0;
    var loadingTime;
    exports.viewContext = { id: id, location: location, sessionId: session.getId() };
    // Update the view every time the measures are changing
    var _a = browser_core_1.throttle(browser_core_1.monitor(updateView), exports.THROTTLE_VIEW_UPDATE_PERIOD, {
        leading: false,
    }), scheduleViewUpdate = _a.throttled, stopScheduleViewUpdate = _a.stop;
    function updateMeasures(newMeasures) {
        measures = tslib_1.__assign(tslib_1.__assign({}, measures), newMeasures);
        scheduleViewUpdate();
    }
    var stopTimingsTracking = trackTimings(lifeCycle, updateMeasures).stop;
    var stopEventCountsTracking = trackEventCounts_1.trackEventCounts(lifeCycle, updateMeasures).stop;
    function updateLoadingTime(loadingTimeValue) {
        loadingTime = loadingTimeValue;
        scheduleViewUpdate();
    }
    var stopLoadingTimeTracking = trackLoadingTime(lifeCycle, loadingType, updateLoadingTime).stop;
    // Initial view update
    updateView();
    function updateView() {
        documentVersion += 1;
        lifeCycle.notify(lifeCycle_1.LifeCycleEventType.VIEW_COLLECTED, {
            documentVersion: documentVersion,
            id: id,
            loadingTime: loadingTime,
            loadingType: loadingType,
            location: location,
            measures: measures,
            duration: performance.now() - startOrigin,
            startTime: startOrigin,
        });
    }
    return {
        end: function () {
            stopTimingsTracking();
            stopEventCountsTracking();
            stopLoadingTimeTracking();
            // prevent pending view updates execution
            stopScheduleViewUpdate();
            // Make a final view update
            updateView();
        },
    };
}
function trackHistory(onHistoryChange) {
    var originalPushState = history.pushState;
    history.pushState = browser_core_1.monitor(function () {
        originalPushState.apply(this, arguments);
        onHistoryChange();
    });
    var originalReplaceState = history.replaceState;
    history.replaceState = browser_core_1.monitor(function () {
        originalReplaceState.apply(this, arguments);
        onHistoryChange();
    });
    window.addEventListener(browser_core_1.DOM_EVENT.POP_STATE, browser_core_1.monitor(onHistoryChange));
}
function areDifferentViews(previous, current) {
    return previous.pathname !== current.pathname;
}
function trackTimings(lifeCycle, callback) {
    var timings = {};
    var stopPerformanceTracking = lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.PERFORMANCE_ENTRY_COLLECTED, function (entry) {
        if (entry.entryType === 'navigation') {
            var navigationEntry = entry;
            timings = tslib_1.__assign(tslib_1.__assign({}, timings), { domComplete: browser_core_1.msToNs(navigationEntry.domComplete), domContentLoaded: browser_core_1.msToNs(navigationEntry.domContentLoadedEventEnd), domInteractive: browser_core_1.msToNs(navigationEntry.domInteractive), loadEventEnd: browser_core_1.msToNs(navigationEntry.loadEventEnd) });
            callback(timings);
        }
        else if (entry.entryType === 'paint' && entry.name === 'first-contentful-paint') {
            var paintEntry = entry;
            timings = tslib_1.__assign(tslib_1.__assign({}, timings), { firstContentfulPaint: browser_core_1.msToNs(paintEntry.startTime) });
            callback(timings);
        }
    }).unsubscribe;
    return { stop: stopPerformanceTracking };
}
function trackLoadingTime(lifeCycle, loadingType, callback) {
    var expectedTiming = 1;
    var receivedTimings = [];
    var stopLoadEventLoadingTime = browser_core_1.noop;
    if (loadingType === ViewLoadingType.INITIAL_LOAD) {
        expectedTiming += 1;
        (stopLoadEventLoadingTime = trackLoadEventLoadingTime(lifeCycle, onTimingValue).stop);
    }
    var stopActivityLoadingTimeTracking = trackActivityLoadingTime(lifeCycle, onTimingValue).stop;
    function onTimingValue(timingValue) {
        expectedTiming -= 1;
        if (timingValue) {
            receivedTimings.push(timingValue);
        }
        if (expectedTiming === 0 && receivedTimings.length) {
            callback(Math.max.apply(Math, receivedTimings));
        }
    }
    return {
        stop: function () {
            stopActivityLoadingTimeTracking();
            stopLoadEventLoadingTime();
        },
    };
}
function trackLoadEventLoadingTime(lifeCycle, callback) {
    var stopPerformanceTracking = lifeCycle.subscribe(lifeCycle_1.LifeCycleEventType.PERFORMANCE_ENTRY_COLLECTED, function (entry) {
        if (entry.entryType === 'navigation') {
            var navigationEntry = entry;
            callback(navigationEntry.loadEventEnd);
        }
    }).unsubscribe;
    return { stop: stopPerformanceTracking };
}
function trackActivityLoadingTime(lifeCycle, callback) {
    var startTime = performance.now();
    var stopWaitIdlePageActivity = trackPageActivities_1.waitIdlePageActivity(lifeCycle, function (hadActivity, endTime) {
        if (hadActivity) {
            callback(endTime - startTime);
        }
        else {
            callback(undefined);
        }
    }).stop;
    return { stop: stopWaitIdlePageActivity };
}
